lib/gateway.js

```diff
@@ -0,0 +1,218 @@
+                  service: serviceMap[serviceForFieldType],
+                  createOperation: createEntityReferenceResolverOperation,
+                  transformData: response => response.json.data._entities[0],
+                  isReference: true
+                })
+              } else {
+                field.resolve = makeResolver({
+                  service: serviceMap[serviceForFieldType],
+                  createOperation: createFieldResolverOperation,
+                  transformData: response => response.json.data._entities[0][fieldName]
+                })
+              }
+            }
+          }
+        } else if (typeFieldsToService[`${type}-${fieldName}`]) {
+          field.resolve = makeResolver({
+            service: serviceMap[typeFieldsToService[`${type}-${fieldName}`]],
+            createOperation: createFieldResolverOperation,
+            transformData: response => response.json.data._entities[0][fieldName]
+          })
+        }
+      }
+    }
+  }
+}
+
+async function buildGateway (gatewayOpts, app) {
+  const { services } = gatewayOpts
+
+  const serviceMap = await buildServiceMap(services)
+
+  const serviceSDLs = Object.values(serviceMap).map(service => service.schemaDefinition)
+
+  const schema = buildFederatedSchema(serviceSDLs.join(''), true)
+
+  const typeToServiceMap = {}
+  const typeFieldsToService = {}
+  const factory = new Factory()
+  app.decorateReply(kEntityResolvels)
+  app.addHook('onRequest', async function (req, reply) {
+    reply[kEntityResolvels] = factory.create()
+  })
+
+  for (const [service, serviceDefinition] of Object.entries(serviceMap)) {
+    for (const type of serviceDefinition.types) {
+      typeToServiceMap[type] = service
+    }
+
+    for (const [type, fields] of Object.entries(serviceDefinition.extensionTypeMap)) {
+      for (const field of fields) {
+        typeFieldsToService[`${type}-${field}`] = service
+      }
+    }
+
+    factory.add(`${service}Entity`, async (queries, context) => {
+      const q = [...new Set(queries.map(q => q.query))]
+
+      const resultIndexes = []
+      let queryIndex = 0
+      const mergedQueries = queries.reduce((acc, curr) => {
+        if (!acc[curr.query]) {
+          acc[curr.query] = curr.variables
+          resultIndexes[q.indexOf(curr.query)] = [queryIndex]
+        } else {
+          acc[curr.query].representations = [
+            ...acc[curr.query].representations,
+            ...curr.variables.representations
+          ]
+          resultIndexes[q.indexOf(curr.query)].push(queryIndex)
+        }
+
+        queryIndex++
+
+        return acc
+      }, {})
+
+      const result = []
+
+      await Promise.all(Object.entries(mergedQueries).map(async ([query, variables], queryIndex, entries) => {
+        const response = await serviceDefinition.sendRequest({
+          method: 'POST',
+          body: JSON.stringify({
+            query,
+            variables
+          })
+        })
```

Comment:
I think you can also combine the queries. It's possible using GraphQL.

Reply:
Added a detailed comment here: https://github.com/mcollina/fastify-gql/pull/130/files#diff-8341b2b42feef9f7f0e004614072f8bdR153

How it is possible, but will be part of a future release